### PR TITLE
Elements are aware of the network

### PR DESCRIPTION
**Pull request by benoit9126**
Summary of the changes:
- `Element` class has a read only field named `network` which is the network it belongs to (may be `None`)
- The `_connect` method of `Element` now accepts a list of elements and the validity of networks of elements to connect is checked.
- The `disconnect` method of `Element` is now called `_disconnect` and handle the `network`
- The `add_element` method of `ElectricalNetwork` is now called `_connect_element`
- The `remove_element` of `ElectricalNetwork` is now called `_disconnect_element`
- A `disconnect` (public) method has been added to the `VoltageSource` and `AbstractLoad` classes to disconnect them from a network. The disconnected object are not usable anymore after their disconnection, as their field `bus` is set to `None`.
- An user warning is raised when accessing a result from any elements of a network if the network is invalid.
- A (single) user warning is raised when accessing any result of an invalid `ElectricalNetwork`
- Tests have been added

### CODE REVIEW

File: roseau/load_flow/models/voltage_sources.py [hunk 1]
```diff
@@ -100,6 +111,10 @@ def from_dict(cls, data: JsonDict) -> "VoltageSource":
         return cls(data["id"], data["bus"], voltages=voltages, phases=data["phases"])
 
     def to_dict(self) -> JsonDict:
+        if self.bus is None:
+            msg = f"The element {self!r} is disconnected and can not be used anymore."
```
Comment (alihamdan) [inline roseau/load_flow/models/voltage_sources.py:115]:
```suggestion
            msg = f"The voltage source {self.id!r} is disconnected and can not be used anymore."
```
Same as the loads

File: roseau/load_flow/models/loads/loads.py [hunk 2]
```diff
@@ -223,15 +231,18 @@ def powers(self) -> np.ndarray:
...
     def res_flexible_powers(self) -> np.ndarray:
         """The load flow result of the load flexible powers (VA)."""
-        if self._res_flexible_powers is None:
-            self._raise_load_flow_not_run()
-        return self._res_flexible_powers
+        return self._res_getter(value=self._res_flexible_powers, warning=True)
 
     def to_dict(self) -> JsonDict:
+        if self.bus is None:
+            msg = f"The element {self!r} is disconnected and can not be used anymore."
```
Comment (alihamdan) [inline roseau/load_flow/models/loads/loads.py:243]:
```suggestion
            msg = f"The load {self.id!r} is disconnected and can not be used anymore."
```
I think this is a better message for the user but no strong opinion.

File: roseau/load_flow/models/loads/loads.py [hunk 3]
```diff
@@ -295,8 +306,13 @@ def currents(self) -> np.ndarray:
     @ureg.wraps(None, (None, "A"), strict=False)
     def currents(self, value: Sequence[complex]) -> None:
         self._currents = self._validate_value(value)
+        self._invalidate_network()
 
     def to_dict(self) -> JsonDict:
+        if self.bus is None:
+            msg = f"The element {self!r} is disconnected and can not be used anymore."
```
Comment (alihamdan) [inline roseau/load_flow/models/loads/loads.py:313]:
```suggestion
            msg = f"The load {self.id!r} is disconnected and can not be used anymore."
```
Same

File: roseau/load_flow/models/loads/loads.py [hunk 4]
```diff
@@ -358,8 +374,13 @@ def impedances(self) -> np.ndarray:
     @ureg.wraps(None, (None, "ohm"), strict=False)
     def impedances(self, impedances: Sequence[complex]) -> None:
         self._impedances = self._validate_value(impedances)
+        self._invalidate_network()
 
     def to_dict(self) -> JsonDict:
+        if self.bus is None:
+            msg = f"The element {self!r} is disconnected and can not be used anymore."
```
Comment (alihamdan) [inline roseau/load_flow/models/loads/loads.py:381]:
```suggestion
            msg = f"The load {self.id!r} is disconnected and can not be used anymore."
```

**Review by alihamdan — COMMENTED** [review]
Looks good. Some minor comments

**Review by Saelyos — APPROVED** [review]

**Review by alihamdan — APPROVED** [review]

**Pull request merged**